### ISSUE TITLE
[SamsungTV] Several Thread Issues

### ISSUE THREAD

**Issue by morph166955** [openhab/openhab-addons#9495]
Opening as per @kaikreuzer from openhab/openhab-core#1998

The SamsungTV binding has almost every method of SamsungTvHandler marked as synchronized, which makes it get blocked when multiple things happen at the same time (here UPnP discovery and value reading).

```
"upnp-main-173" #5582 prio=5 os_prio=0 cpu=52.86ms elapsed=1184.08s tid=0x00007fbd78154800 nid=0x43ac1 waiting for monitor entry  [0x00007fbcf4eff000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:69)
	- waiting to lock <0x00000000c452b620> (a java.lang.Object)
	at org.openhab.binding.samsungtv.internal.protocol.RemoteControllerWebSocket.openConnection(RemoteControllerWebSocket.java:147)
	at org.openhab.binding.samsungtv.internal.service.RemoteControllerService.start(RemoteControllerService.java:203)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.checkCreateManualConnection(SamsungTvHandler.java:312)
	- locked <0x00000000c452ca20> (a org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.checkAndCreateServices(SamsungTvHandler.java:253)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.thingDiscovered(SamsungTvHandler.java:362)
	at org.openhab.core.config.discovery.internal.DiscoveryServiceRegistryImpl$1.run(DiscoveryServiceRegistryImpl.java:260)

"OH-thingHandler-1" #141 daemon prio=5 os_prio=0 cpu=5576.05ms elapsed=3419.65s tid=0x00007fbd20052800 nid=0x40c8a waiting for monitor entry  [0x00007fbd05d76000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.valueReceived(SamsungTvHandler.java:225)
	- waiting to lock <0x00000000c452ca20> (a org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler)
	at org.openhab.binding.samsungtv.internal.service.MediaRendererService.onValueReceived(MediaRendererService.java:205)
	at org.openhab.binding.samsungtv.internal.service.MediaRendererService.updateResourceState(MediaRendererService.java:240)
	at org.openhab.binding.samsungtv.internal.service.MediaRendererService.handleCommand(MediaRendererService.java:121)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.poll(SamsungTvHandler.java:217)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler$$Lambda$1161/0x0000000101137440.run(Unknown Source)
```

Additionally, it registers as a discovery listener (also not really allowed for a handler) and creates websocket connections to the remote device instead of returning the call immediately, which blocks the whole discovery infrastructure:

```
"upnp-main-173" #5582 prio=5 os_prio=0 cpu=52.86ms elapsed=1184.08s tid=0x00007fbd78154800 nid=0x43ac1 waiting for monitor entry  [0x00007fbcf4eff000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:69)
	- waiting to lock <0x00000000c452b620> (a java.lang.Object)
	at org.openhab.binding.samsungtv.internal.protocol.RemoteControllerWebSocket.openConnection(RemoteControllerWebSocket.java:147)
	at org.openhab.binding.samsungtv.internal.service.RemoteControllerService.start(RemoteControllerService.java:203)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.checkCreateManualConnection(SamsungTvHandler.java:312)
	- locked <0x00000000c452ca20> (a org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.checkAndCreateServices(SamsungTvHandler.java:253)
	at org.openhab.binding.samsungtv.internal.handler.SamsungTvHandler.thingDiscovered(SamsungTvHandler.java:362)
	at org.openhab.core.config.discovery.internal.DiscoveryServiceRegistryImpl$1.run(DiscoveryServiceRegistryImpl.java:260
```

**Comment by arjanmels** [comment]
Thanks for figuring this out. Anyone planning to fix this. Otherwise I may have some time to work on it in the next couple of weeks (but no certainty).

**Comment by Hilbrand** [comment]
This looks like the problem reported in #8419. There was also a pr to fix this #7499. But it was never reported that it fixed the problem as no one tested it or at least didn't report it. Then the pr got auto closed in the main rename and I'm not sure if @lolodomo has any interest in picking this up again... Or if he agrees someone can pick that code up and update it for openHAB 3.

**Comment by kaikreuzer** [comment]
Thanks the for the links @Hilbrand! I have created https://github.com/openhab/openhab-addons/pull/9525 and volunteer to drive this forward (just bought a SamsungTV for x-mas, so after setting it up, I should hopefully be also able to test it).

**Comment by morph166955** [comment]
@kai welcome to the hell of automating SamsungTVs...  They definitely don't make it easy. I'm happy to test a jar as needed here, I've got some older TVs as well as one of the first to use the websockets.

**Comment by jaywiseman1971** [comment]
@kai,  I'm also wanting the Samsung binding to be fixed since earlier this year.  I'm still running OH 2.5 so I'm would like to test some jar's also if the plan is to fix it at the 2.5 level also?

Best, Jay

**Comment by Cossey** [comment]
@kaikreuzer awesome! Will you consider back-porting the fixes to 2.5?￼￼ 

Just one thing to note, the way it detects the web socket TVs being online is through UPnP, but that causes ￼issues when it needs to power on via either Wake on LAN or sending the remote code "power" through the web socket. When the TV turns off it takes awhile for the web socket server to shut down and when it is still running you can only send the remote code to turn it on, not the WoL packet. ￼￼￼
